api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,268 @@ package messagebuffer
 
 import (
 	"encoding/hex"
+	"fmt"
+	"io"
 	"sync"
+	"time"
 
+	rbt "github.com/emirpasic/gods/trees/redblacktree"
+	"github.com/nknorg/nkn/v2/config"
 	"github.com/nknorg/nkn/v2/pb"
+	"github.com/nknorg/nkn/v2/util/log"
 )
 
+const maxSeqId int64 = 1000000 // it should be less than 1 million messages per milli-second
+const MonitorInterval = 10     // seconds
+
+// msg node with create time and expire time
+type msgNode struct {
+	msg      *pb.Relay
+	createId int64
+	expireId int64
+}
+
+func (m *msgNode) GetSize() int {
+	// m.createId(int64), m.expireId(int64), m.msg.MaxHoldingSeconds(int32), m.msg.SigChainLen(int32)
+	return (8 + 8 + 4 + 4 + len(m.msg.SrcIdentifier) + len(m.msg.SrcPubkey) + len(m.msg.DestId) +
+		len(m.msg.Payload) + len(m.msg.BlockHash) + len(m.msg.LastHash))
+}
+
+type indexNode struct {
+	clientID string
+	createId int64
+	expireId int64
+}
+
 // MessageBuffer is the buffer to hold message for clients not online
+// here use several locks, must use locks in the same order to avoid dead lock.
+// the order should be: msgMu -> createMu -> expireMu
 type MessageBuffer struct {
 	sync.Mutex
-	buffer map[string][]*pb.Relay
+	buffer     map[string]*rbt.Tree // map ClinetId to msgs tree
+	treeCreate *rbt.Tree            // index tree of create time
+	treeExpire *rbt.Tree            // index tree of expire time
+
+	totalBufSize int   // the buffer size in bytes
+	seqId        int64 // sequence id
+}
+
+// comparation function of red and black tree
+func compare(a, b interface{}) (res int) {
+	aInt := a.(int64)
+	bInt := b.(int64)
+	switch {
+	case aInt < bInt:
+		res = -1
+	case aInt == bInt:
+		res = 0
+	case aInt > bInt:
+		res = 1
+	}
+	return
 }
 
 // NewMessageBuffer creates a MessageBuffer
 func NewMessageBuffer() *MessageBuffer {
-	return &MessageBuffer{
-		buffer: make(map[string][]*pb.Relay),
+
+	msgBuffer := &MessageBuffer{
+		buffer:     make(map[string]*rbt.Tree), // *pb.Relay
+		treeCreate: rbt.NewWith(compare),
+		treeExpire: rbt.NewWith(compare),
 	}
+
+	go msgBuffer.BufferMonitor()
+
+	return msgBuffer
 }
 
 // AddMessage adds a message to message buffer
-func (messageBuffer *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
+func (msgBuf *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
+
+	if msg.MaxHoldingSeconds <= 0 { // no need cache
+		return
+	}
+
 	clientIDStr := hex.EncodeToString(clientID)
-	messageBuffer.Lock()
-	defer messageBuffer.Unlock()
-	messageBuffer.buffer[clientIDStr] = append(messageBuffer.buffer[clientIDStr], msg)
+	now := time.Now().UnixMilli()
+
+	msgBuf.Lock()
+	defer msgBuf.Unlock()
+
+	msgBuf.seqId++
+	if msgBuf.seqId >= maxSeqId {
```

Comment:
If this reset is triggered, message order might be change. An example:

```
receive msg 1
receive msg 2
receive msg 3
seqId reset
receive msg 4
receive msg 5
receive msg 6
```

But messages will be delivered in order `4 5 6 1 2 3`

I think we can use this solution:
keep track of last value of `now`. If current `now` is different from last `now`, or if `msgBuf.seqId >= maxSeqId`, we trigger seq id reset.  This way, reset is only triggered when there are more than a million messages in the same millisecond, which will not happen practically.

Reply:
yes.  modified.